config/read.go

```diff
@@ -771,20 +781,19 @@ func Read(logger *util.Logger, filename string) (Config, error) {
 				SystemScope: config.SystemScope,
 			}
 
-			if config.GetDbName() != "" {
-				// Ensure we have no duplicate identifiers within one collector
-				skip := false
-				for _, server := range conf.Servers {
-					if config.Identifier == server.Identifier {
-						skip = true
-					}
-				}
-				if skip {
-					logger.PrintError("Skipping config section %s, detected as duplicate. Note: To monitor multiple databases on the same server, db_name accepts a comma-separated list.", config.SectionName)
-				} else {
-					conf.Servers = append(conf.Servers, *config)
+			// Ensure we have no duplicate identifiers within one collector
+			skip := false
+			for _, server := range conf.Servers {
+				if config.Identifier == server.Identifier {
+					skip = true
 				}
 			}
+			if skip {
+				logger.PrintError("Skipping config section %s, detected as duplicate. Note: To monitor multiple databases on the same server, db_name accepts a comma-separated list.", config.SectionName)
+				continue
+			}
+
+			conf.Servers = append(conf.Servers, *config)
 
 			if config.DbURL != "" {
```

Comment:
It's a little weird that we do this check _after_ we already process and append the server to the server list. I didn't touch it, but I was tempted to. Thoughts?

Reply:
Oh also, we only log an error there, instead of returning an error. I don't understand why.

Reply:
Ask Maciek 4 years ago?
https://github.com/pganalyze/collector/commit/ac45be47c7b525a9fad9928e3398dba71f332319

> Validate at startup, but allow to continue to avoid retrofitting error
handling across helpers that right now cannot return errors. The helpers
return dummy/default values which will fail authentication.

Though honestly reading these, I still don't understand well for why we can't return the error here.

Reply:
Oh man, that guy. He writes the worst code. And the guy who approved that: he'll approve anything!

I'm pretty sure the reason it ended up being a `PrintError` message originally is that the error above, about duplicate config sections, is handled with `PrintError`. On reflection, (1) that case is a little different, so not returning an error makes a little more sense and (2) it still does not make enough sense to do it that way, and we should change the duplicate section handling to return an error.

But I think changing duplicate section handling could break some currently-working configurations, so I'm not super excited about doing that. Whereas in this case, if you don't have a valid DB URL, you're not going to get very far with the collector, so I think it's safe to error out earlier. I'll update the code to do that.

Reply:
Okay, updated.